pkg/services/store/resolver/ds_cache.go

```diff
@@ -0,0 +1,135 @@
+package resolver
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/grafana/grafana/pkg/plugins/manager/registry"
+	"github.com/grafana/grafana/pkg/services/datasources"
+	"github.com/grafana/grafana/pkg/services/store"
+	"github.com/grafana/grafana/pkg/tsdb/grafanads"
+)
+
+type dsVal struct {
+	InternalID   int64
+	IsDefault    bool
+	Name         string
+	Type         string
+	UID          string
+	PluginExists bool // type exists
+}
+
+type dsCache struct {
+	ds             datasources.DataSourceService
+	pluginRegistry registry.Service
+	cache          map[int64]map[string]*dsVal
+	timestamp      time.Time // across all orgIDs
+	mu             sync.Mutex
+}
```

Comment:
don't we have a cache for data sources? https://github.com/grafana/grafana/blob/main/pkg/services/datasources/datasources.go#L60

Reply:
we do... but I am not sure using it helps much...  the approach in the PR lets you resolve by name/id/uid, just like the frontend.   

Using the existing one would require somehow encoding orgId -- not sure how to best do that (in the user context?).  Actually when I look at the interface [here](https://github.com/grafana/grafana/blob/main/pkg/services/datasources/datasources.go#L65) I would bet big money that the code assumes UID is unique across orgs! 😬

Reply:
Existing cache gets orgId though the user object and sets it as `uidKey(user.OrgID, datasourceUID)`

Maybe we can use that... but then look-up everything when we need to do it by name?  once everything is upgraded, that could be OK, but until then I'm not sure if it is a good pattern